crates/astria-conductor/src/sequencer.rs

```diff
@@ -0,0 +1,340 @@
+            None => bail!("subscription to sequencer for new blocks failed immediately; bailing"),
+            Some(Err(e)) => {
+                return Err(e).wrap_err("first sequencer block returned from subscription was bad");
+            }
+            Some(Ok(block)) => {
+                let height = block.header().height.value();
+                pending_blocks.push_back(futures::future::ready(Ok(block)));
+                height
+            }
+        };
+
+        let latest_height: u32 = latest_height.try_into().wrap_err(
+            "failed converting the cometbft height to u32, but this should always work",
+        )?;
+
+        info!(
+            height.initial = self.initial_sequencer_block_height,
+            height.latest = latest_height,
+            "syncing sequencer between configured initial and latest retrieved height"
+        );
+        let mut sync_heights = {
+            let sync_range = self.initial_sequencer_block_height..latest_height;
+            let pool = self.pool.clone();
+            Box::pin(futures::stream::iter(sync_range).then(move |height| {
+                let pool = pool.clone();
+                async move { pool.get().await }.map_ok(move |client| (height, client))
+            }))
+        }
+        .fuse();
+
+        let mut sync_blocks = FuturesOrdered::new();
+
+        let mut resubscribe = future::Fuse::terminated();
+        'reader_loop: loop {
+            select! {
+                shutdown = &mut self.shutdown => {
+                    let ret = match shutdown {
+                        Err(e) => {
+                            warn!(error.message = %e, "shutdown channel closed unexpectedly; shutting down");
+                            Err(e).wrap_err("shut down channel closed unexpectedly")
+                        }
+                        Ok(()) => {
+                            info!("received shutdown signal; shutting down");
+                            Ok(())
+                        }
+                    };
+                    break 'reader_loop ret;
+                }
+
+                // Drain the stream of heights to sync the sequencer reader.
+                // The stream is fused, so after it exhausted this match arm will be deactivated.
+                // The condition on the length of priority blocks relies on the pool being currently set
+                // to 50. This ensures that no more than 20 requests to the sequencer are active at the same time.
+                // Leaving some objects in the pool is important, so that the match arm for resolving priority blocks
+                // below can reschedule blocks.
+                Some(res) = sync_heights.next(), if !sync_heights.is_done() && sync_blocks.len() < 20 => {
+                    if let Err(e) = sync_block_at_height(&mut sync_blocks, res) {
+                        error!(error.message = %e, error.cause = ?e, "syncing heights failed; exiting reader");
```

Comment:
if there is an error grabbing a client it seems like this will trigger full reader loop exit?

Reply:
Yes, exactly. Assume the driver connection failed, then `pool.get().await` will essentially block on `WebSocketClient::new(url).await` in `ClientProvider` resolving, so the steps are:

1. call `pool.get().await`
2. create a one-shot channel, the sender half is sent to `ClientProvider` 
3. in `get()`: immediately block on `rx.await` to get a client from `ClientProvider`
4. in ClientProvider: if `client = None`, push the `tx` half into `pending_requests`
5. wait for the reconnection to resolve
6. if successful: drain `pending_requests`, sending a new client to each requester
7. if failed: exit the loop, dropping `pending_requests`, which drops all `tx`s closing the channel
8. in `get()`: `recv.await` will only fail if the tx has been dropped